pkg/queue/concurrency_state.go

```diff
@@ -38,50 +40,48 @@ func ConcurrencyStateHandler(logger *zap.SugaredLogger, h http.Handler, pause, r
 		resume = func() {}
 	}
 
-	type req struct {
-		w http.ResponseWriter
-		r *http.Request
+	var (
+		inFlight = atomic.NewInt64(0)
+		paused   = true
+		mux      sync.RWMutex
+	)
 
-		done chan struct{}
-	}
-
-	reqCh := make(chan req)
-	doneCh := make(chan struct{})
-	go func() {
-		inFlight := 0
-
-		// This loop is entirely synchronous, so there's no cleverness needed in
-		// ensuring open and close dont run at the same time etc. Only the
-		// delegated ServeHTTP is done in a goroutine.
-		for {
-			select {
-			case <-doneCh:
-				inFlight--
-				if inFlight == 0 {
-					logger.Info("Requests dropped to zero ...")
+	return func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if inFlight.Dec() == 0 {
+				mux.Lock()
+				defer mux.Unlock()
+				// We need to doublecheck this since another request can have reached the
+				// handler meanwhile. We don't want to do anything in that case.
+				if !paused && inFlight.Load() == 0 {
 					pause()
+					paused = true
 				}
+			}
+		}()
 
-			case r := <-reqCh:
-				inFlight++
-				if inFlight == 1 {
-					logger.Info("Requests increased from zero ...")
-					resume()
-				}
+		inFlight.Inc()
 
-				go func(r req) {
-					h.ServeHTTP(r.w, r.r)
-					close(r.done) // Return from ServeHTTP
-					doneCh <- struct{}{}
-				}(r)
-			}
+		mux.RLock()
+		if !paused {
+			// General stable-state case.
+			defer mux.RUnlock()
+			h.ServeHTTP(w, r)
+			return
+		}
+		mux.RUnlock()
+		mux.Lock()
+		if !paused { // doubly-checked locking
+			// Another request raced us and resumed already.
```

Comment:
So here do this again is only for performance？Have any other benifits?

Reply:
Yes, https://en.wikipedia.org/wiki/Double-checked_locking. Doubly checked locking has the benefit that only a read lock is required for the "common" case. We're assuming that the vast majority will take the "// General stable-state case." route above, thus only fetching a read-lock, which allows for concurrent execution.